File: sdks/go/crypto.go
package hsm

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/sha512"
	"crypto/subtle"
	"encoding/base64"
	"encoding/hex"
	"regexp"
	"strings"
)

// ToBase64 encodes bytes to base64.
func ToBase64(data []byte) string {
	return base64.StdEncoding.EncodeToString(data)
}

// FromBase64 decodes base64 to bytes.
func FromBase64(data string) ([]byte, error) {
	return base64.StdEncoding.DecodeString(data)
}

// IsBase64 checks if a string is valid base64.
func IsBase64(s string) bool {
	if s == "" {
		return true
	}
	if len(s)%4 != 0 {
		return false
	}
	pattern := regexp.MustCompile(`^[A-Za-z0-9+/]*={0,2}$`)
	return pattern.MatchString(s)
}

// ToHex encodes bytes to hex.
func ToHex(data []byte) string {
	return hex.EncodeToString(data)
}

// FromHex decodes hex to bytes.
func FromHex(data string) ([]byte, error) {
	data = strings.TrimPrefix(data, "0x")
	return hex.DecodeString(data)
}

// NormalizeToBase64 normalizes data to base64 for API requests.
func NormalizeToBase64(data []byte) string {
	return ToBase64(data)
}

// NormalizeStringToBase64 normalizes a string to base64.
func NormalizeStringToBase64(data string) string {
	if IsBase64(data) {
		return data
	}
	return ToBase64([]byte(data))
}

// SHA256 hashes data using SHA-256.
func SHA256(data []byte) []byte {
	hash := sha256.Sum256(data)
	return hash[:]
}

// SHA384 hashes data using SHA-384.
func SHA384(data []byte) []byte {
	hash := sha512.Sum384(data)
	return hash[:]
}

// SHA512 hashes data using SHA-512.
func SHA512Hash(data []byte) []byte {
	hash := sha512.Sum512(data)
	return hash[:]
}

// HMACSHA256 computes HMAC-SHA256.
func HMACSHA256(key, data []byte) []byte {
	h := hmac.New(sha256.New, key)
	h.Write(data)
	return h.Sum(nil)
}

// ConstantTimeEqual compares two byte slices in constant time.
func ConstantTimeEqual(a, b []byte) bool {
	return subtle.ConstantTimeCompare(a, b) == 1
}

// RandomBytes generates cryptographically secure random bytes.
func RandomBytes(n int) ([]byte, error) {
	bytes := make([]byte, n)
	_, err := rand.Read(bytes)
	return bytes, err
}

// ConcatBytes concatenates multiple byte slices.
func ConcatBytes(slices ...[]byte) []byte {
	totalLen := 0
	for _, s := range slices {
		totalLen += len(s)
	}
	result := make([]byte, 0, totalLen)
	for _, s := range slices {
		result = append(result, s...)
	}
	return result
}

// DERToRaw converts signature format from DER to raw (R || S).
func DERToRaw(derSignature []byte) ([]byte, error) {
	if len(derSignature) == 0 || derSignature[0] != 0x30 {
		// Already in raw format or invalid
		return derSignature, nil
	}

	offset := 2 // Skip 0x30 and length byte

	// Parse R
	if derSignature[offset] != 0x02 {
		return nil, NewValidationError("Invalid DER signature", "")
	}
	offset++
	rLen := int(derSignature[offset])
	offset++
	r := derSignature[offset : offset+rLen]
	offset += rLen

	// Parse S
	if derSignature[offset] != 0x02 {
		return nil, NewValidationError("Invalid DER signature", "")
	}
	offset++
	sLen := int(derSignature[offset])
	offset++
	s := derSignature[offset : offset+sLen]

	// Remove leading zeros if present
	if len(r) > 32 && r[0] == 0x00 {
		r = r[1:]
	}
	if len(s) > 32 && s[0] == 0x00 {
		s = s[1:]
	}

	// Pad to 32 bytes each
	rPadded := make([]byte, 32)
	sPadded := make([]byte, 32)
	copy(rPadded[32-len(r):], r)
	copy(sPadded[32-len(s):], s)

	return ConcatBytes(rPadded, sPadded), nil
}

// RawToDER converts signature format from raw (R || S) to DER.
func RawToDER(rawSignature []byte) ([]byte, error) {
	if len(rawSignature) != 64 {
		// Might already be DER or invalid
		return rawSignature, nil
	}

	r := rawSignature[:32]
	s := rawSignature[32:64]

	// Remove leading zeros
	for len(r) > 1 && r[0] == 0x00 && r[1]&0x80 == 0 {
		r = r[1:]
	}
	for len(s) > 1 && s[0] == 0x00 && s[1]&0x80 == 0 {
		s = s[1:]
	}

	// Add leading zero if high bit is set
	if r[0]&0x80 != 0 {
		r = append([]byte{0x00}, r...)
	}
	if s[0]&0x80 != 0 {
		s = append([]byte{0x00}, s...)
	}

	totalLen := 4 + len(r) + len(s)

	result := make([]byte, 0, 2+totalLen)
	result = append(result, 0x30, byte(totalLen))
	result = append(result, 0x02, byte(len(r)))
	result = append(result, r...)
	result = append(result, 0x02, byte(len(s)))
	result = append(result, s...)

	return result, nil
}

File: sdks/typescript/src/crypto.ts
/**
 * HSM Crypto Helpers
 *
 * Utility functions for cryptographic operations.
 */

/**
 * Encode bytes to base64
 */
export function toBase64(data: Uint8Array | string): string {
  if (typeof data === 'string') {
    // Already base64 or text - assume text and encode
    if (isBase64(data)) {
      return data;
    }
    data = new TextEncoder().encode(data);
  }

  // Node.js environment
  if (typeof Buffer !== 'undefined') {
    return Buffer.from(data).toString('base64');
  }

  // Browser environment
  let binary = '';
  for (let i = 0; i < data.length; i++) {
    binary += String.fromCharCode(data[i]);
  }
  return btoa(binary);
}

/**
 * Decode base64 to bytes
 */
export function fromBase64(data: string): Uint8Array {
  // Node.js environment
  if (typeof Buffer !== 'undefined') {
    return new Uint8Array(Buffer.from(data, 'base64'));
  }

  // Browser environment
  const binary = atob(data);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

/**
 * Check if string is valid base64
 */
export function isBase64(str: string): boolean {
  if (str.length === 0) return true;
  const base64Regex = /^[A-Za-z0-9+/]*={0,2}$/;
  return base64Regex.test(str) && str.length % 4 === 0;
}

/**
 * Encode bytes to hex
 */
export function toHex(data: Uint8Array): string {
  return Array.from(data)
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

/**
 * Decode hex to bytes
 */
export function fromHex(hex: string): Uint8Array {
  const cleanHex = hex.startsWith('0x') ? hex.slice(2) : hex;
  const bytes = new Uint8Array(cleanHex.length / 2);
  for (let i = 0; i < cleanHex.length; i += 2) {
    bytes[i / 2] = parseInt(cleanHex.slice(i, i + 2), 16);
  }
  return bytes;
}

/**
 * Normalize data to base64 for API requests
 */
export function normalizeToBase64(data: string | Uint8Array): string {
  if (typeof data === 'string') {
    // If already base64, return as-is
    if (isBase64(data)) {
      return data;
    }
    // Otherwise, encode as UTF-8 text
    return toBase64(new TextEncoder().encode(data));
  }
  return toBase64(data);
}

/**
 * Hash data using SHA-256 (browser-compatible)
 */
export async function sha256(data: Uint8Array | string): Promise<Uint8Array> {
  const bytes = typeof data === 'string' ? new TextEncoder().encode(data) : data;

  // Use Web Crypto API (available in both browser and Node.js 20+)
  if (typeof crypto !== 'undefined' && crypto.subtle) {
    const hashBuffer = await crypto.subtle.digest('SHA-256', bytes);
    return new Uint8Array(hashBuffer);
  }

  // Fallback for older Node.js
  if (typeof require !== 'undefined') {
    const cryptoModule = require('crypto');
    return new Uint8Array(cryptoModule.createHash('sha256').update(bytes).digest());
  }

  throw new Error('No crypto implementation available');
}

/**
 * Hash data using Keccak-256 (Ethereum)
 * Note: For full Ethereum support, use a dedicated library like ethers.js
 */
export async function keccak256(data: Uint8Array | string): Promise<Uint8Array> {
  // This is a placeholder - for production use, implement or import keccak256
  // For now, we'll use SHA-256 as a fallback with a warning
  console.warn('keccak256: Using SHA-256 as fallback. For Ethereum operations, use ethers.js or similar.');
  return sha256(data);
}

/**
 * Create a message hash for signing (pre-hash if needed)
 */
export async function createMessageHash(
  message: string | Uint8Array,
  algorithm: 'SHA-256' | 'SHA-384' | 'SHA-512' | 'KECCAK-256' = 'SHA-256'
): Promise<Uint8Array> {
  const bytes = typeof message === 'string' ? new TextEncoder().encode(message) : message;

  if (algorithm === 'KECCAK-256') {
    return keccak256(bytes);
  }

  if (typeof crypto !== 'undefined' && crypto.subtle) {
    const hashBuffer = await crypto.subtle.digest(algorithm, bytes);
    return new Uint8Array(hashBuffer);
  }

  if (typeof require !== 'undefined') {
    const cryptoModule = require('crypto');
    const nodeAlgorithm = algorithm.toLowerCase().replace('-', '');
    return new Uint8Array(cryptoModule.createHash(nodeAlgorithm).update(bytes).digest());
  }

  throw new Error('No crypto implementation available');
}

/**
 * Compare two byte arrays in constant time
 */
export function constantTimeEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) {
    return false;
  }

  let result = 0;
  for (let i = 0; i < a.length; i++) {
    result |= a[i] ^ b[i];
  }
  return result === 0;
}

/**
 * Generate random bytes
 */
export function randomBytes(length: number): Uint8Array {
  const bytes = new Uint8Array(length);

  if (typeof crypto !== 'undefined' && crypto.getRandomValues) {
    crypto.getRandomValues(bytes);
    return bytes;
  }

  if (typeof require !== 'undefined') {
    const cryptoModule = require('crypto');
    return new Uint8Array(cryptoModule.randomBytes(length));
  }

  throw new Error('No crypto implementation available');
}

/**
 * Concatenate multiple byte arrays
 */
export function concatBytes(...arrays: Uint8Array[]): Uint8Array {
  const totalLength = arrays.reduce((sum, arr) => sum + arr.length, 0);
  const result = new Uint8Array(totalLength);
  let offset = 0;
  for (const arr of arrays) {
    result.set(arr, offset);
    offset += arr.length;
  }
  return result;
}

/**
 * Split bytes at a given index
 */
export function splitBytes(data: Uint8Array, index: number): [Uint8Array, Uint8Array] {
  return [data.slice(0, index), data.slice(index)];
}

/**
 * Convert signature format from DER to raw (R || S)
 * Useful for Ethereum and other chains that expect raw format
 */
export function derToRaw(derSignature: Uint8Array): Uint8Array {
  // Simple DER parsing - for production, use a proper ASN.1 library
  // DER format: 0x30 [total-len] 0x02 [r-len] [r] 0x02 [s-len] [s]

  if (derSignature[0] !== 0x30) {
    // Already in raw format or invalid
    return derSignature;
  }

  let offset = 2; // Skip 0x30 and length byte

  // Parse R
  if (derSignature[offset] !== 0x02) {
    throw new Error('Invalid DER signature');
  }
  offset++;
  const rLen = derSignature[offset];
  offset++;
  let r = derSignature.slice(offset, offset + rLen);
  offset += rLen;

  // Parse S
  if (derSignature[offset] !== 0x02) {
    throw new Error('Invalid DER signature');
  }
  offset++;
  const sLen = derSignature[offset];
  offset++;
  let s = derSignature.slice(offset, offset + sLen);

  // Remove leading zeros if present (DER encoding adds them for positive numbers)
  if (r[0] === 0x00 && r.length > 32) {
    r = r.slice(1);
  }
  if (s[0] === 0x00 && s.length > 32) {
    s = s.slice(1);
  }

  // Pad to 32 bytes each
  const rPadded = new Uint8Array(32);
  const sPadded = new Uint8Array(32);
  rPadded.set(r, 32 - r.length);
  sPadded.set(s, 32 - s.length);

  return concatBytes(rPadded, sPadded);
}

/**
 * Convert signature format from raw (R || S) to DER
 */
export function rawToDer(rawSignature: Uint8Array): Uint8Array {
  if (rawSignature.length !== 64) {
    // Might already be DER or invalid
    return rawSignature;
  }

  let r = rawSignature.slice(0, 32);
  let s = rawSignature.slice(32, 64);

  // Remove leading zeros
  while (r.length > 1 && r[0] === 0x00 && !(r[1] & 0x80)) {
    r = r.slice(1);
  }
  while (s.length > 1 && s[0] === 0x00 && !(s[1] & 0x80)) {
    s = s.slice(1);
  }

  // Add leading zero if high bit is set (to ensure positive integer)
  if (r[0] & 0x80) {
    r = concatBytes(new Uint8Array([0x00]), r);
  }
  if (s[0] & 0x80) {
    s = concatBytes(new Uint8Array([0x00]), s);
  }

  const totalLen = 4 + r.length + s.length; // 2 bytes for each INTEGER header

  return new Uint8Array([
    0x30, // SEQUENCE
    totalLen,
    0x02, // INTEGER (R)
    r.length,
    ...r,
    0x02, // INTEGER (S)
    s.length,
    ...s,
  ]);
}
